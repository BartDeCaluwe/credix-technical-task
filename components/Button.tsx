import React from 'react'

interface ButtonProps {
  /**
   * Button contents
   */
  label: string
  /**
   * Optional click handler
   */
  onClick?: () => void
}

/**
 * Primary UI component for user interaction
 */
export const Button = ({ label, ...props }: ButtonProps) => {
  return (
    <button
      type="button"
      className="px-6 py-2 bg-teal-700 text-white"
      {...props}
    >
      {label}
    </button>
  )
}
