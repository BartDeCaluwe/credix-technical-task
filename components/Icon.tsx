export type IconNames = 'bar' | 'calendar' | 'link'
interface IconProps {
  /**
   * Name of the icon to be rendered
   */
  name: IconNames

  /**
   * Optional width of the icon
   */
  width?: number

  /**
   * Optional height of the icon
   */
  height?: number
}

export const Icon = ({ name, width, height }: IconProps) => {
  switch (name) {
    case 'bar':
      return (
        <svg
          className={`w-${width} h-${height}`}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
        >
          <path d="M12 8a1 1 0 0 0-1 1v11a1 1 0 0 0 2 0V9a1 1 0 0 0-1-1z" />
          <path d="M19 4a1 1 0 0 0-1 1v15a1 1 0 0 0 2 0V5a1 1 0 0 0-1-1z" />
          <path d="M5 12a1 1 0 0 0-1 1v7a1 1 0 0 0 2 0v-7a1 1 0 0 0-1-1z" />
        </svg>
      )
    case 'calendar':
      return (
        <svg
          className={`w-${width} h-${height}`}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
        >
          <g data-name="Layer 2">
            <g data-name="calendar">
              <rect width="24" height="24" opacity="0" />
              <path d="M18 4h-1V3a1 1 0 0 0-2 0v1H9V3a1 1 0 0 0-2 0v1H6a3 3 0 0 0-3 3v12a3 3 0 0 0 3 3h12a3 3 0 0 0 3-3V7a3 3 0 0 0-3-3zM6 6h1v1a1 1 0 0 0 2 0V6h6v1a1 1 0 0 0 2 0V6h1a1 1 0 0 1 1 1v4H5V7a1 1 0 0 1 1-1zm12 14H6a1 1 0 0 1-1-1v-6h14v6a1 1 0 0 1-1 1z" />
              <circle cx="8" cy="16" r="1" />
              <path d="M16 15h-4a1 1 0 0 0 0 2h4a1 1 0 0 0 0-2z" />
            </g>
          </g>
        </svg>
      )
    case 'link':
      return (
        <svg
          className={`w-${width} h-${height}`}
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          strokeWidth="2"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71" />
          <path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71" />
        </svg>
      )

    default:
      return null
  }
}
