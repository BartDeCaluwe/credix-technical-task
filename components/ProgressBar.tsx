import { useClamp } from '../hooks/useClamp'

interface ProgressBarProps {
  /**
   * The actual progress value as a percentage between 0 and 100
   */
  value: number
}

export const ProgressBar = ({ value }: ProgressBarProps) => {
  const clamped = useClamp(value, 0, 100)

  return (
    <div>
      <div className="text-sm mb-1">{clamped}%</div>
      <div className="relative min-w-[100px]">
        <div className="bg-gray-300 h-[4px] w-full"></div>
        <div
          className={[
            'absolute top-0 left-0 h-[4px] flex justify-end items-center',
            clamped === 0 ? 'bg-gray-300' : 'bg-teal-700',
            `w-[${clamped}%]`,
          ].join(' ')}
        >
          <span
            className={[
              'bg-teal-700 flex-shrink-0 inline-block h-2 w-2 rounded-full',
              clamped === 0 ? '' : '-mr-1 ',
            ].join(' ')}
            aria-hidden="true"
          ></span>
        </div>
      </div>
    </div>
  )
}
