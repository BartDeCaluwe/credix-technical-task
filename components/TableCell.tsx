import { ReactNode } from 'react'

interface TableCellProps {
  /**
   * Optional children to be rendered inside the component
   */
  children?: ReactNode

  /**
   * Optional className to add or override classes
   */
  className?: string

  /**
   * Optional table cell alignment specifier
   */
  align?: 'left' | 'right' | 'center'
}

export const TableCell = ({
  children,
  className,
  ...props
}: TableCellProps) => {
  return (
    <td className={['px-6 py-4', className].join(' ')} {...props}>
      {children}
    </td>
  )
}
