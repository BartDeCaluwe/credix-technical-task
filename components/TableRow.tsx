import { ReactNode } from 'react'

interface TableRowProps {
  /**
   * Optional children to be rendered inside the component
   */
  children?: ReactNode
}

export const TableRow = ({ children }: TableRowProps) => {
  return <tr className="bg-white">{children}</tr>
}
