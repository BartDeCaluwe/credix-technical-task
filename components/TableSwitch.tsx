interface TableSwitchProps {
  isActive: boolean
  label: string
  onClick?: () => void
}

export const TableSwitch = ({
  isActive,
  label,
  ...props
}: TableSwitchProps) => {
  const borderColor = isActive ? 'border-green-800' : 'border-white'

  return (
    <div
      className={[
        'bg-white px-6 py-4 w-32 min-w-min border-b-4 flex justify-center hover:cursor-pointer hover:bg-gray-50',
        borderColor,
      ].join(' ')}
      {...props}
    >
      {label}
    </div>
  )
}
