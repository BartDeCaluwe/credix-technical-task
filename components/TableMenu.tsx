import { FC } from 'react'
import { Icon } from './Icon'

export interface TableMenuItem {
  /**
   * Table menu item contents
   */
  label: string

  /**
   * Optional boolean to make the label invisible but still accessible for screen readers
   */
  isInvisible?: boolean

  /**
   * Optional icon
   */
  icon?: string

  /**
   * Optional alignment specifier
   */
  align?: 'start' | 'end' | 'center'
}
interface TableMenuProps {
  /**
   * Collection of menu items
   */
  items: TableMenuItem[]
}

export const TableMenu = ({ items, ...props }: TableMenuProps) => {
  return (
    <thead className="bg-white" {...props}>
      {items.map(({ label, icon, isInvisible, align, ...itemProps }) => (
        <th
          key={`table-menu-${label}`}
          className="px-6 py-4 text-md font-medium"
          {...itemProps}
        >
          <div
            className={[
              'flex gap-1 items-center',
              isInvisible ? 'invisible' : 'visible',
              `justify-${align}`,
            ].join(' ')}
          >
            {icon && <Icon name={icon} width={4} height={4} />}
            {label}
          </div>
        </th>
      ))}
    </thead>
  )
}
