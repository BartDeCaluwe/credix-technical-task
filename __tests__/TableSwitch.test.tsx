import React from 'react'
import renderer from 'react-test-renderer'
import { TableSwitch } from '../components/TableSwitch'

test('TableRow renders active state', () => {
  const component = renderer.create(
    <TableSwitch isActive={true} label="Table switch" />
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TableRow renders inactive state', () => {
  const component = renderer.create(
    <TableSwitch isActive={false} label="Table switch" />
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TableRow has a min width', () => {
  const component = renderer.create(<TableSwitch isActive={false} label="T" />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
