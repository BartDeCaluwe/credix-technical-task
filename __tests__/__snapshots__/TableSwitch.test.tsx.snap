// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow has a min width 1`] = `
<button
  className="bg-white px-6 py-4 w-32 min-w-min border-b-4 flex justify-center hover:cursor-pointer hover:bg-gray-50 border-white"
  type="button"
>
  T
</button>
`;

exports[`TableRow renders active state 1`] = `
<button
  className="bg-white px-6 py-4 w-32 min-w-min border-b-4 flex justify-center hover:cursor-pointer hover:bg-gray-50 border-teal-800"
  type="button"
>
  Table switch
</button>
`;

exports[`TableRow renders inactive state 1`] = `
<button
  className="bg-white px-6 py-4 w-32 min-w-min border-b-4 flex justify-center hover:cursor-pointer hover:bg-gray-50 border-white"
  type="button"
>
  Table switch
</button>
`;
